test/core/xds/xds_listener_resource_type_test.cc

```diff
@@ -0,0 +1,1962 @@
+        "  }\n"
+        "}");
+    if (!bootstrap.ok()) {
+      gpr_log(GPR_ERROR, "Error parsing bootstrap: %s",
+              bootstrap.status().ToString().c_str());
+      GPR_ASSERT(false);
+    }
+    return MakeRefCounted<XdsClient>(std::move(*bootstrap),
+                                     /*transport_factory=*/nullptr);
+  }
+
+  RefCountedPtr<XdsClient> xds_client_;
+  upb::DefPool upb_def_pool_;
+  upb::Arena upb_arena_;
+  XdsResourceType::DecodeContext decode_context_;
+};
+
+TEST_F(XdsListenerTest, Definition) {
+  auto* resource_type = XdsListenerResourceType::Get();
+  ASSERT_NE(resource_type, nullptr);
+  EXPECT_EQ(resource_type->type_url(), "envoy.config.listener.v3.Listener");
+  EXPECT_TRUE(resource_type->AllResourcesRequiredInSotW());
+}
+
+TEST_F(XdsListenerTest, UnparseableProto) {
+  std::string serialized_resource("\0", 1);
+  auto* resource_type = XdsListenerResourceType::Get();
+  auto decode_result =
+      resource_type->Decode(decode_context_, serialized_resource);
+  EXPECT_EQ(decode_result.resource.status().code(),
+            absl::StatusCode::kInvalidArgument);
+  EXPECT_EQ(decode_result.resource.status().message(),
+            "Can't parse Listener resource.")
+      << decode_result.resource.status();
+}
+
+TEST_F(XdsListenerTest, NeitherAddressNotApiListener) {
+  Listener listener;
+  listener.set_name("foo");
+  std::string serialized_resource;
+  ASSERT_TRUE(listener.SerializeToString(&serialized_resource));
+  auto* resource_type = XdsListenerResourceType::Get();
+  auto decode_result =
+      resource_type->Decode(decode_context_, serialized_resource);
+  EXPECT_EQ(decode_result.resource.status().code(),
+            absl::StatusCode::kInvalidArgument);
+  EXPECT_EQ(decode_result.resource.status().message(),
+            "Listener has neither address nor ApiListener")
+      << decode_result.resource.status();
+}
+
+// TODO(roth): Re-enable the following test once
+// github.com/istio/istio/issues/38914 is resolved.
+TEST_F(XdsListenerTest, DISABLED_BothAddressAndApiListener) {
+  Listener listener;
+  listener.set_name("foo");
+  listener.mutable_api_listener();
+  listener.mutable_address();
+  std::string serialized_resource;
+  ASSERT_TRUE(listener.SerializeToString(&serialized_resource));
+  auto* resource_type = XdsListenerResourceType::Get();
+  auto decode_result =
+      resource_type->Decode(decode_context_, serialized_resource);
+  EXPECT_EQ(decode_result.resource.status().code(),
+            absl::StatusCode::kInvalidArgument);
+  EXPECT_EQ(decode_result.resource.status().message(),
+            "Listener has both address and ApiListener")
+      << decode_result.resource.status();
+}
+
+//
+// HttpConnectionManager tests
+//
+
+// These tests cover common behavior for both API listeners and TCP
+// listeners, so we run them in both contexts.
+class HttpConnectionManagerTest : public XdsListenerTest,
+                                  public ::testing::WithParamInterface<bool> {
```

Comment:
nit: It might be better to have a class/struct instead of a raw bool

Reply:
Done.